text/0000-router-service.md

```diff
@@ -0,0 +1,440 @@
+
+### New Methods: Initiating Transitions
+
+```js
+transitionTo(routeName, ...models, queryParams)
+replaceWith(routeName, ...models, queryParams)
+```
+
+These two have the same semantics as the existing methods on `Ember.Route`:
+
+### New Method: Checking For Active Route
+
+ - `isActive(routeName, ...models, queryParams)`
+
+The arguments have the same semantics as `transitionTo`, the return value is a boolean. This should provide the same logic that determines whether to put an active class on a `link-to`. Here's an example of how we can implement `is-active` as a helper, using this method:
+
+```js
+import Helper from 'ember-helper';
+import service from 'ember-service/inject';
+import observer from 'ember-metal/observer';
+
+export default Helper.extend({
+  router: service(),
+  compute([routeName, ...models], hash) {
+    let allModels;
+    if (hash.models) {
+      allModels = models.concat(hash.models);
+    } else {
+      allModels = models;
+    }
+    return this.get('router').isActive(routeName, ...models, hash.queryParams);
+  },
+  didTransition: observer('router.currentRoute', function() {
+    this.recompute();
+  });
+});
+```
+
+```hbs
+{{!- Example usage -}}
+<li class={{if (is-active "person.detail" model) 'chosen'}} >
+
+{{!- Example usage with generic routeName and list of models (avoids splat) -}}
+<a class={{if (is-active routeName models=models) 'chosen'}} >
+
+{{!- Note that the complexities of currentWhen can be avoided by composing instead. }}
+<a class={{if (or (is-active 'one') (is-active 'two')) 'active'}} href={{url-for 'two'}} >
+
+```
+
+### New Method: URL generation
+
+`urlFor(routeName, ...models, queryParams)`
+
+This takes the same arguments as `transitionTo`, but instead of initiating the transition it returns the resulting root-relative URL as a string (which will include the application's `rootUrl`).
+
+A `url-for` helper can be implemented almost identically to the `is-active` example above.
+
+### New Method: URL recognition
+
+`recognize(url)`
+
+Takes a string URL and returns a `RouteInfo` for the leafmost route represented by the URL. Returns `null` if the URL is not recognized. This method expects to receive the actual URL as seen by the browser _including_ the app's `rootURL`.
+
+
+Example: this feature can replace [this use of private API in ember-href-to](https://github.com/intercom/ember-href-to/blob/b8cf0699eec6a65570b05e4fc22b27d8cea49c42/app/instance-initializers/browser/ember-href-to.js#L34).
+
+
+### New Method: Recognize and load models
+
+`recognizeAndLoad(url)`
+
+Takes a string URL and returns a promise that resolves to a `RouteInfoWithAttributes` for the leafmost route represented by the URL. The promise rejects if the URL is not recognized or an unhandled exception is encountered. This method expects to receive the actual URL as seen by the browser _including_ the app's `rootURL`.
```

Comment:
It seems like this must process the entire hierarchy's "model chain" in order to guarantee safety. I'm not sure how you can just load the leaf? We get around this by creating a separate `prefetch` hook which knows it isn't allowed to depend on anything unless it explicitly wires up its own promise chain. Nonetheless it still relies on all the `prefetch` hooks in the hierarchy being invoked in order so that the user can manually wire together a promise chain.